File: src/frontend/src/lib/permissions.ts
export type PermissionKey = string;

export const permissions = {
  dashboardView: "dashboard.view",
  auditView: "audit.view",
  usersView: "users.view",
  usersCreate: "users.create",
  usersEdit: "users.edit",
  usersActivate: "users.activate",
  usersResetPassword: "users.reset_password",
  usersAssignRoles: "users.assign_roles",
  rolesView: "roles.view",
  rolesCreate: "roles.create",
  rolesEdit: "roles.edit",
  rolesDelete: "roles.delete",
  rolesAssignPermissions: "roles.assign_permissions",
  productsView: "products.view",
  productsCreate: "products.create",
  productsEdit: "products.edit",
  productsDelete: "products.delete",
  productsImport: "products.import",
  vendorsView: "vendors.view",
  vendorsManage: "vendors.manage",
  categoriesView: "categories.view",
  categoriesManage: "categories.manage",
  templatesView: "templates.view",
  templatesCreate: "templates.create",
  templatesEdit: "templates.edit",
  templatesPreview: "templates.preview",
  templatesCompare: "templates.compare",
  templatesSubmitReview: "templates.submit_review",
  templatesReview: "templates.review",
  templatesPublish: "templates.publish",
  templatesArchive: "templates.archive",
  templatesRestore: "templates.restore",
  assetsView: "assets.view",
  assetsUpload: "assets.upload",
  assetsDelete: "assets.delete",
  printIntentsView: "printintents.view",
  printIntentsCreate: "printintents.create",
  printIntentsHandoff: "printintents.handoff",
  printIntentsDispatch: "printintents.dispatch",
  printIntentsConfirm: "printintents.confirm",
  printIntentsFail: "printintents.fail",
  printIntentsCancel: "printintents.cancel",
} as const;

const legacyRolePermissionMap: Record<string, PermissionKey[]> = {
  Admin: [],
  Operator: [
    permissions.productsCreate,
    permissions.productsEdit,
    permissions.productsImport,
    permissions.vendorsManage,
    permissions.categoriesManage,
    permissions.templatesCreate,
    permissions.templatesEdit,
    permissions.templatesSubmitReview,
    permissions.templatesRestore,
    permissions.assetsView,
    permissions.assetsUpload,
    permissions.printIntentsCreate,
    permissions.printIntentsHandoff,
    permissions.printIntentsDispatch,
    permissions.printIntentsConfirm,
    permissions.printIntentsFail,
    permissions.printIntentsCancel,
  ],
  Reviewer: [
    permissions.templatesReview,
    permissions.templatesPublish,
    permissions.templatesRestore,
  ],
  Viewer: [],
};

export function hasPermission(userPermissions: string[] | undefined, requiredPermission: PermissionKey): boolean {
  return (userPermissions || []).includes(requiredPermission);
}

export function hasAnyPermission(userPermissions: string[] | undefined, requiredPermissions: PermissionKey[]): boolean {
  const granted = new Set(userPermissions || []);
  return requiredPermissions.some((permission) => granted.has(permission));
}

export function hasLegacyRoleAccess(userRoles: string[] | undefined, userPermissions: string[] | undefined, allowedRoles: string[]): boolean {
  const roles = userRoles || [];
  const permissionSet = new Set(userPermissions || []);

  return allowedRoles.some((allowedRole) => {
    if (allowedRole === "Viewer") {
      return true;
    }

    if (roles.includes(allowedRole)) {
      return true;
    }

    const mappedPermissions = legacyRolePermissionMap[allowedRole];
    if (!mappedPermissions) {
      return false;
    }

    if (allowedRole === "Admin") {
      return roles.includes("Admin");
    }

    return mappedPermissions.some((permission) => permissionSet.has(permission));
  });
}

File: src/backend/Plms.Api/Security/Permissions.cs
namespace Plms.Api.Security
{
    public static class Permissions
    {
        public const string DashboardView = "dashboard.view";
        public const string AuditView = "audit.view";

        public const string UsersView = "users.view";
        public const string UsersCreate = "users.create";
        public const string UsersEdit = "users.edit";
        public const string UsersActivate = "users.activate";
        public const string UsersResetPassword = "users.reset_password";
        public const string UsersAssignRoles = "users.assign_roles";

        public const string RolesView = "roles.view";
        public const string RolesCreate = "roles.create";
        public const string RolesEdit = "roles.edit";
        public const string RolesDelete = "roles.delete";
        public const string RolesAssignPermissions = "roles.assign_permissions";

        public const string ProductsView = "products.view";
        public const string ProductsCreate = "products.create";
        public const string ProductsEdit = "products.edit";
        public const string ProductsDelete = "products.delete";
        public const string ProductsImport = "products.import";

        public const string VendorsView = "vendors.view";
        public const string VendorsManage = "vendors.manage";

        public const string CategoriesView = "categories.view";
        public const string CategoriesManage = "categories.manage";

        public const string TemplatesView = "templates.view";
        public const string TemplatesCreate = "templates.create";
        public const string TemplatesEdit = "templates.edit";
        public const string TemplatesPreview = "templates.preview";
        public const string TemplatesCompare = "templates.compare";
        public const string TemplatesSubmitReview = "templates.submit_review";
        public const string TemplatesReview = "templates.review";
        public const string TemplatesPublish = "templates.publish";
        public const string TemplatesArchive = "templates.archive";
        public const string TemplatesRestore = "templates.restore";
        public const string TemplatesDelete = "templates.delete";

        public const string AssetsView = "assets.view";
        public const string AssetsUpload = "assets.upload";
        public const string AssetsDelete = "assets.delete";

        public const string PrintIntentsView = "printintents.view";
        public const string PrintIntentsCreate = "printintents.create";
        public const string PrintIntentsHandoff = "printintents.handoff";
        public const string PrintIntentsDispatch = "printintents.dispatch";
        public const string PrintIntentsConfirm = "printintents.confirm";
        public const string PrintIntentsFail = "printintents.fail";
        public const string PrintIntentsCancel = "printintents.cancel";
    }
}
